risc0/zkvm/platform/src/syscall.rs

```diff
@@ -541,59 +583,58 @@ pub unsafe extern "C" fn sys_getenv(
-            // heap.
-            //
-            // This is generated automatically by the linker; see
-            // https://lld.llvm.org/ELF/linker_script.html#sections-command
-            static _end: u8;
-        }
-
-        // Pointer to next heap address to use, or 0 if the heap has not yet been
-        // initialized.
-        static mut HEAP_POS: usize = 0;
+pub extern "C" fn sys_alloc_aligned(bytes: usize, align: usize) -> *mut u8 {
+    extern "C" {
+        // This symbol is defined by the loader and marks the end
+        // of all elf sections, so this is where we start our
+        // heap.
+        //
+        // This is generated automatically by the linker; see
+        // https://lld.llvm.org/ELF/linker_script.html#sections-command
+        static _end: u8;
+    }
 
-        // SAFETY: Single threaded, so nothing else can touch this while we're working.
-        let mut heap_pos = unsafe { HEAP_POS };
+    // Pointer to next heap address to use, or 0 if the heap has not yet been
+    // initialized.
+    static mut HEAP_POS: usize = 0;
 
-        if heap_pos == 0 {
-            heap_pos = (&_end) as *const u8 as usize;
-        }
+    // SAFETY: Single threaded, so nothing else can touch this while we're working.
+    let mut heap_pos = unsafe { HEAP_POS };
 
-        let offset = heap_pos & (align - 1);
-        if offset != 0 {
-            heap_pos += align - offset;
-        }
+    if heap_pos == 0 {
+        heap_pos = unsafe { (&_end) as *const u8 as usize };
+    }
 
-        // Ensure all allocations are minimally aligned to a word boundary.
-        let align = usize::min(align, WORD_SIZE);
+    let offset = heap_pos & (align - 1);
+    if offset != 0 {
+        heap_pos += align - offset;
+    }
 
-        let ptr = heap_pos as *mut u8;
-        heap_pos += bytes;
+    // Ensure all allocations are minimally aligned to a word boundary.
+    let align = usize::min(align, WORD_SIZE);
 
-        // Check to make sure we keep space between the heap and the
-        // stack so they don't accidentally step on each other.
-        let mut stack_pointer: usize;
-        unsafe { asm!("add {stack_pointer}, sp, zero", stack_pointer = out(reg) stack_pointer) };
-        if stack_pointer - (crate::memory::RESERVED_STACK as usize) < heap_pos {
-            const MSG: &[u8] = "Out of memory!".as_bytes();
-            sys_panic(MSG.as_ptr(), MSG.len());
-        }
+    let ptr = heap_pos as *mut u8;
+    heap_pos += bytes;
 
-        unsafe { HEAP_POS = heap_pos };
-        ptr
+    // Check to make sure we keep space between the heap and the
+    // stack so they don't accidentally step on each other.
+    let mut stack_pointer: usize;
+    #[cfg(target_os = "zkvm")]
+    unsafe {
+        asm!("add {stack_pointer}, sp, zero", stack_pointer = out(reg) stack_pointer)
+    };
+    #[cfg(not(target_os = "zkvm"))]
+    let stack_pointer: usize = crate::memory::STACK_TOP as usize;
```

Comment:
I see that this is now enabled for `#[cfg(not(target_os = "zkvm"))]` what are the use cases for this function outside of the zkvm?

Reply:
This was simply to make tooling (like clippy and rust-analyzer) aware of more code. It's not meant to ever be executed except on the `zkvm` target.

Reply:
ok thanks for the context!